src/libstd/io/mod.rs

```diff
@@ -116,19 +92,45 @@ fn append_to_string<F>(buf: &mut String, f: F) -> Result<usize>
     }
 }
 
+// This uses an adaptive system to extend the vector when it fills. We want to
+// avoid paying to allocate and zero a huge chunk of memory if the reader only
+// has 4 bytes while still making large reads if the reader does have a ton
+// of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
+// time is 4,500 times (!) slower than this if the reader has a very small
+// amount of data to return.
 fn read_to_end<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> Result<usize> {
-    let mut read = 0;
+    let start_len = buf.len();
+    let mut len = start_len;
+    let mut cap_bump = 16;
+    let ret;
     loop {
-        if buf.capacity() == buf.len() {
-            buf.reserve(DEFAULT_BUF_SIZE);
+        if len == buf.len() {
+            if buf.capacity() == buf.len() {
+                if cap_bump < DEFAULT_BUF_SIZE {
```

Comment:
Looking at the implementation of reserve() I think there's two separate issues.
1. It looks like reserve() actually always rounds up to the nearest power of two. Because of that, the algorithm is basically equivalent to my suggestion of always doubling the capacity of the vector, and so is O(number of bytes). However this is not guaranteed by the reserve() API and thus a change to reserve, holding the API consistent, could make this algorithm O(n^2) (or maybe worse).
-if the exact behavior of reserve is meant to be part of the API, and will be included in the API, then this algorithm can be made equivalent by always reserving 16.
-if the behavior of reserve is never intended to be part of the API, then this should probably be coded here directly with reserve_exact to make sure the algorithm is always O(number of bytes)

2.
> As is, that's 32k of wasted work, but if there was no cap it could be 10s or 100s of megabytes
Not so. Let's pretend we start with a vector that has 32k capacity and size, and need to read in 64 megabytes + 1 byte (so we are doing the maximum wasted work for the doubling algorithm).  Let's assume that copying (ie a realloc) and memsetting are equally expensive.
For the doubling algorithm, we need to copy and memset on (32k + 64 k + 128k + 256k + 512k + ... + ... + 64 MB + 128 MB) = 128(1 + 1/2 + 1/4 + ...) MB ~= 256 MB.
For the increase by 32k algorithm, we need to copy and memset on (32k + 64k + 96k + 128k + 160k + ... + (64 MB - 32k) + (64 MB) + (64 MB + 32k) ~= 32(1 + 2 + 3 + ... + 2000 + 2001) kilobytes = 32 * (2001 * 2002) / 2 kilobytes ~= 64e6 kilobytes = 64 gigabytes(!)
Being O(n^2), this behavior gets even worse for larger amounts of data, and 64 MB is not at all of an unusual amount of data to be working with.